google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentReference.java

```diff
@@ -45,7 +45,7 @@
 public class DocumentReference {
 
   private final ResourcePath path;
-  private final FirestoreImpl firestore;
+  private static FirestoreImpl firestore;
```

Comment:
why `static?` and why not `static final` ?

Reply:
I dont think this would be correct solution, GCP clients does not prefer static clientimpl fields.

Reply:
> why `static?` and why not `static final` ?


because static method required static data member.
if i used static final it will ask me for initialization if i initialize  it with  null 
ie.private static final FirestoreImpl firestore = null it will throw an error in DocumentReference.
 
 protected DocumentReference(
      FirestoreImpl firestore, ResourcePath path) { // Elevated access level for mocking.
    this.path = path;
    this.firestore = firestore;//getting error here cannot assign value to final variable 'firestore'
  }

Reply:
A user never asked for static `FirestoreImpl`. He is expecting it to get public( which I don't think static is needed for that).

Reply:
> Note, that currently it requires FirestoreImpl which is package-private.
>
> If possible, a static method that converts a Document to java object would be preferable.


he expects only a static method for conversion of Document to java object.

Reply:
User wants to create document snapshot objcet from the document without firestoreimpl.currently what happened if you want to used DocumentSnapshot.toObject(...) You also have to passed firestoreimpl right.without firestoreimpl User is not able to directly used DocumentSnapshot.toObject(...) Right?

Reply:
Is `Document` is a proto class?.. if yes then there will be an equivalent java class will be there.

> currently what happened if you want to used DocumentSnapshot.toObject(...) You also have to passed firestoreimpl

Then why to change this behavior, he shall get an instance of FirestoreImpl by new builder or constructor, then he will use this instance and create Document using the static method.

Reply:
> Is `Document` is a proto class?.. if yes then there will be an equivalent java class will be there.
> 
> > currently what happened if you want to used DocumentSnapshot.toObject(...) You also have to passed firestoreimpl
> 
> Then why to change this behavior, he shall get an instance of FirestoreImpl by new builder or constructor, then he will use this instance and create Document using the static method.

Intentionally User don't want to use firestoreimpl